Split preference store

idafurjes opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Splitting the preferences sqlstore into Store and Service. Store handles CRUD operations and Service additional logic.

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Ref #46522

**Special notes for your reviewer**:

Commit:
Split preference store

zserge commented on pkg/services/preference/pref.go:
```diff
@@ -0,0 +1,11 @@
+type Service interface {
+	GetPreferenceWithDefaults(context.Context, *GetPreferenceWithDefaultsQuery) (*Preference, error)
+	GetPreference(context.Context, *GetPreferenceQuery) (*Preference, error)
+	SavePreference(context.Context, *SavePreferenceCommand) (*Preference, error)
```
Hm... why does it have to return preference instance?

zserge commented on pkg/services/preference/prefimpl/pref_store.go:
```diff
@@ -0,0 +1,119 @@
+	return &prefs, err
+}
+
+func (s *sqlStore) GetDefaults() *pref.Preference {
```
I don't think this method should be part of the store at all

zserge replied:
Also, I believe removing it will disconnect store from the cfg

idafurjes replied:
It doesn't belong to store, I can move it to the service or somewhere else?

yangkb09 committed:
Chore: Add tests to pref

Commit:
Fix preference in wire

Commit:
Rename and adjust

Commit:
Add pref service test

Commit:
Rename methods, add tests

Commit:
Merge branch 'main' into ida/kat/pref

Commit:
Rename Preferences to Preference, names IDs correctly

Commit:
Merge branch 'main' into ida/kat/pref

Commit:
Fix lint

Commit:
Refactor Save

Commit:
Refactor upsert
Add new logic for QueryHistory
Rename some fields according to go naming conventions
Refactore tests

Commit:
Merge branch 'main' into ida/kat/pref

Commit:
Roll back ID that breaks tests

Commit:
Rename Id to ID in UpdatePreferenceQuery

zserge commented:
I mostly like it, but I wonder if the store API can be improved to use a single _model_ struct and maybe reusing it for get/insert/update?

I realise that for this very store we probably wouldn't have a second implementation, but if we theoretically imagine some object/blob store - these xorm-specific and sql-specific commands would make no sense at the store API/interface level.

idafurjes commented:
I understand, it would be easy to remove Insert/Update, but I decided to keep it this way because of the Get struct.

Commit:
Use preference as a model to modify store

sakjur commented on pkg/services/preference/prefimpl/pref.go:
```diff
@@ -0,0 +1,203 @@
+	if exists {
+		err = s.store.Update(ctx, preference)
+	} else {
+		_, err = s.store.Insert(ctx, preference)
```
Is `Insert` something a `Patch` operation should be able to do?

idafurjes replied:
I kept the naming and the functionality from sqlstore

zserge replied:
I agree that at the Store level we should try to keep names closer to CRUD, since most of the possible Store implementations use that, while Patch is very unusual (i.e. S3 doesn't allow you to patch an object). Maybe even narrow it down to Put/Get/Delete/List. But I'm fine with the current naming, too.

idafurjes replied:
this is the endpoint PATCH /api/user/preferences, maybe we can rename on the PR where we implement the service in the api

sakjur reviewed (approved):
:+1: It'd be nice to get this merged so that we can take the next step of moving over references to Preferences in SQLStore to instead use this service 🙂

I started trying the PR locally before remembering that we're not actually hooking it up to the API yet 😅 It would be nice to get to try that, especially if that allowed us to update this service to support `HomeDashboardUID` without having to also update all the legacy stuff.

idafurjes commented on pkg/services/preference/preftest/fake.go:
```diff
@@ -0,0 +1,60 @@
+func (f *FakePreferenceService) GetDefaults() *pref.Preference {
+	return f.ExpectedPreference
+}
+
+type FakePreferenceStore struct {
+	ExpectedPreference      *pref.Preference
+	ExpectedListPreferences []*pref.Preference
+	ExpectedID              int64
+	ExpectedError           error
+	ExpectedGetError        error
+}
+
+func NewPreferenceStoreFake() *FakePreferenceStore {
+	return &FakePreferenceStore{}
+}
+
+func (f *FakePreferenceStore) List(ctx context.Context, query *pref.Preference) ([]*pref.Preference, error) {
+	return f.ExpectedListPreferences, f.ExpectedError
+}
+
+func (f *FakePreferenceStore) Get(ctx context.Context, query *pref.Preference) (*pref.Preference, error) {
+	return f.ExpectedPreference, f.ExpectedGetError
+}
+
+func (f *FakePreferenceStore) Insert(ctx context.Context, cmd *pref.Preference) (int64, error) {
+	return f.ExpectedID, f.ExpectedError
+}
+
+func (f *FakePreferenceStore) Update(ctx context.Context, cmd *pref.Preference) error {
+	return f.ExpectedError
+}
```
you mean just to keep it in the test file?

zserge commented on pkg/services/preference/model.go:
```diff
@@ -0,0 +1,100 @@
+	OrgID           int64   `xorm:"org_id"`
+	UserID          int64   `xorm:"user_id"`
+	TeamID          int64   `xorm:"team_id"`
+	Teams           []int64 `xorm:"extends"`
```
Yes, XORM does lots of magic under the hood anyway, and I'd rather keep that clearly expressed as struct tags (akin to JSON), than let it be deduced from the struct field names, forcing us to name things as "Id" or name struct types in a certain manner. Once we agree to move away from XORM - removing tags would be easy, but renaming fields and structs back would take time.

papagian replied:
😮  I didn't know about `extends`

idafurjes replied:
I just read the docs you shared with me :)

Commit:
Move pref store fakes to pref test file

papagian commented on pkg/services/preference/prefimpl/store_test.go:
```diff
@@ -0,0 +1,165 @@
+		require.Equal(t, int64(1), prefs[0].HomeDashboardID)
+	})
+
+	t.Run("Update for a user should only modify a single value", func(t *testing.T) {
```
this fails with PostgreSQL for some reason:
```
=== RUN   TestPreferencesDataAccess/Update_for_a_user_should_only_modify_a_single_value
    store_test.go:155: Result mismatch (-want +got):
          &pref.Preference{
                ... // 7 identical fields
                Timezone:  "browser",
                WeekStart: "1",
        -       Theme:     "light",
        +       Theme:     "dark",
                Created:   s"2022-04-14 08:27:05 +0000 UTC",
                Updated:   s"2022-04-14 08:27:05 +0000 UTC",
                JSONData:  &{Navbar: {SavedItems: {{ID: "alerting", Text: "Alerting", Url: "/alerting"}}}},
          }

I run it using:
```
GRAFANA_TEST_DB=postgres go test -v ./pkg/services/preference/prefimpl/... -tags=integration -run TestPreferencesDataAccess
```

idafurjes replied:
oh yes, thanks Sofia

Commit:
Add integration tag for store tests

Commit:
Adjust test

ivanahuckova reviewed (commented):
Changes related to query history preferences look good! 🙂
